[devtools/visualization] Add visualize_graph

Erik-Lundell opened the pull request:
When working with passes, you might have access to a modified graph_module rather than an exported_program. visualize_graph allows visualization of this graph_module by combining  the modified graph_module with an exported_program. Note that the graph_module can't be set directly, a new exported_program needs to be constructed.

Additionally, we disable the operator validation for the newly constructed ExportedProgram. This is ok since it is only used for visualization.

Erik-Lundell committed:
[devtools/visualization] Add visualize_graph

When working with passes, you might have access to
a modified graph_module rather than an exported_program.
visualize_graph allows visualization of this graph_module
by combining  the modified graph_module with an exported_program.
Note that the graph_module can't be set directly, a new
exported_program needs to be constructed.

Additionally, we disable the operator validation for the
newly constructed ExportedProgram. This is ok since
it is only used for visualization.

Signed-off-by: Erik Lundell <erik.lundell@arm.com>
Change-Id: I4fad809bf094a1ec70e25534cc0858f9d8d3d225

zingo commented:
Hi @digantdesai this PR touch non Arm folders, OK?

digantdesai commented on devtools/visualization/visualization_utils.py:
```diff
@@ -124,3 +128,29 @@ def visualize(
     )
+
+
+def visualize_graph(
```
Curious, why not overload the `visualize` method, and add support for additional graph_module in `_get_exported_program()`?

Erik-Lundell replied:
I would argue that constructing a new ExportedProgram from an old + GraphModule is different enough from just visualizing that it's good to signal that with another function name.

digantdesai commented on devtools/visualization/visualization_utils_test.py:
```diff
@@ -145,6 +147,17 @@ def test_visualize_to_executorch(server):
+            exported_program
+        ).exported_program()
+        modified_gm = DecomposeLinearPass()(exported_program.graph_module).graph_module
+        visualize_graph(modified_gm, exported_program)
```
For the future, IMO it is more valuable to show a "diff view" (not sure how it will work) when a graph module is provided along with an exported program? Because most likely that is what I would care about.

Erik-Lundell replied:
That's a very interesting idea! Definitely something to think about for the future.

digantdesai commented on devtools/visualization/visualization_utils_test.py:
```diff
@@ -158,3 +171,5 @@ def test_visualize_to_executorch(server):
         test_visualize_partition(SingletonModelExplorerServer)
         test_visualize_to_executorch(SingletonModelExplorerServer)
+        test_visualize_graph(SingletonModelExplorerServer)
+        time.sleep(3.0)
```
Comment why?

Erik-Lundell replied:
Ack

digantdesai commented on devtools/visualization/visualization_utils_test.py:
```diff
@@ -145,6 +147,17 @@ def test_visualize_to_executorch(server):
+        exported_program = to_edge_transform_and_lower(
+            exported_program
+        ).exported_program()
+        modified_gm = DecomposeLinearPass()(exported_program.graph_module).graph_module
```
why can't we just `_transform` and create a new ep and use `visualize(new_ep)`?

Erik-Lundell replied:
One use case I had in mind for this was inside passes. Also, if you use your own PassManager, it returns a PassResult that contains a graph_module. So in this particular test case you could do that but I think this approach is more flexible.

Erik-Lundell committed:
Use default overloads when calling custom ops

If a node is created without specifying an overload,
A OpOverloadPacket is created, rather than an OpOverload.
This works in a GraphModule, but the OpOverloadPacket is not a valid
operator type in the _EXIREdgeDialectVerifier, which means
that Edge ExportedPrograms can't contain a GraphModule with
such ops.

In short, specifying using the default overload seems
to be the more correct way of creating a custom
operator.

Signed-off-by: Erik Lundell <erik.lundell@arm.com>
Change-Id: I3a1733c0ae88826d88b1e820eaacff765df7fbd2
